compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,456 @@
+            for ty in tys.iter() {
+                s.push_str(&encode_ty(tcx, ty, dict, encode_c_types));
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Array(ty0, len) => {
+            // A<array-length><element-type>
+            let mut s = String::from("A");
+            s.push_str(&format!("{}", &len.val().try_to_scalar().unwrap().to_i64().unwrap()));
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Slice(ty0) => {
+            // u5sliceI<element-type>E as vendor extended type
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        ty::Adt(adt_def, substs) => {
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.opt_item_name(def_id).unwrap());
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+            // respective builtin type encoding as specified by the Itanium C++ ABI if
+            // encode_c_types is true.
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                // C char, integer, and void types
+                s.push('v');
+            } else if adt_def.repr().c() {
+                // <length><name>
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
+                compress(dict, (ty, TyQ::None), &mut s);
+            } else if adt_def.is_struct() && adt_def.repr().transparent() {
+                // <element-type1>
+                let mut tys = adt_def.all_fields().map(|field| field.ty(tcx, substs));
+                let ty0 = tys.next().unwrap();
+                s.push_str(&encode_ty(tcx, ty0, dict, encode_c_types));
+            } else {
+                // u<length><type-name>[I<element-type1..element-typeN>E] as vendor extended type
+
+                // Start and namespace tags
+                let mut def_path = tcx.def_path(def_id);
+                def_path.data.reverse();
+                for disambiguated_data in &def_path.data {
+                    s.push_str(match disambiguated_data.data {
+                        hir::definitions::DefPathData::TypeNs(_) => "Nt",
+                        hir::definitions::DefPathData::ValueNs(_) => "Nv",
+                        _ => {
+                            bug!("encode_ty: unexpected `{:?}`", disambiguated_data.data);
+                        },
```

Comment:
Is any other start and namespace tag possible here?

Reply:
<details><summary>ClosureExpr, AnonConsts, maybe others?</summary>

```rust
#![feature(inline_const)]
#![feature(type_alias_impl_trait)]

type T1 = impl Send;
type T2 = impl Send;
type T3 = impl Send;

pub fn test<T>() -> T {
    todo!();
}

pub fn f() {
    || {
        pub struct A;
        let _: T1 = A;
    };
    const {
        pub struct A;
        fn g() -> T2 { A }
    };

    pub struct A(u32);
    let _: T3 = A;

    test::<T1>();
    test::<T2>();
    test::<T3>();
}
```
</details>

Reply:
Done. I'm handling the same as v0. However, I'm not sure about ForeignMod. Is it actually possible here? Or is it possible in `ty::Foreign` only?

Reply:
See comment above (tl;dr: it's used for`ty::Foreign` only, so I changed it back to `bug!`, since it's not possible in `encode_ty_name()` because it's not used for `ty::Foreign`).